include/rmm/device_buffer.hpp

```diff
@@ -236,6 +236,39 @@ class device_buffer {
     _stream = cuda_stream_view{};
   }
 
+  /**
+   * @brief Increase the capacity of the device memory allocation
+   *
+   * If the requested `new_capacity` is less than or equal to `capacity()`, no
+   * action is taken.
+   *
+   * If `new_capacity` is larger than `capacity()`, a new allocation is made on
+   * `stream` to satisfy `new_capacity`, and the contents of the old allocation are
+   * copied on `stream` to the new allocation. The old allocation is then freed.
+   * The bytes from `[size(), new_capacity)` are uninitialized.
+   *
+   * @throws rmm::bad_alloc If creating the new allocation fails
+   * @throws rmm::cuda_error if the copy from the old to new allocation
+   * fails
+   *
+   * @param new_capacity The requested new capacity, in bytes
+   * @param stream The stream to use for allocation and copy
+   */
+  void reserve(std::size_t new_capacity, cuda_stream_view stream)
+  {
+    set_stream(stream);
+    if (new_capacity > capacity()) {
+      void* const new_data = _mr->allocate(new_capacity, this->stream());
+      auto const old_size  = size();
+      RMM_CUDA_TRY(
+        cudaMemcpyAsync(new_data, data(), size(), cudaMemcpyDefault, this->stream().value()));
+      deallocate_async();
+      _data     = new_data;
+      _size     = old_size;
+      _capacity = new_capacity;
```

Comment:
Could this be cleaner with something like the copy and swap idiom?  Would need to define `swap`.

```suggestion
    if (new_capacity > capacity()) {
     auto new_buffer = device_buffer(new_capacity, stream);
     auto old_size = size();
     RMM_CUDA_TRY(
        cudaMemcpyAsync(new_buffer.data(), data(), old_size(), cudaMemcpyDefault, this->stream().value()));
     swap(new_buffer, *this);
     _size     = old_size;
    }
```

Reply:
Sounds sensible! I'll also make the same change in `resize`, if that is okay to include in this PR